tools/chpldef/compute-goto-entity.cpp

```diff
@@ -0,0 +1,263 @@
+  int l1 = pos.line + 1;
+  int c1 = pos.character + 1;
+  int l2 = pos.line + 1;
+  int c2 = pos.character + 1;
+  return chpl::Location(path, l1, c1, l2, c2);
+}
+
+static const chpl::uast::AstNode*
+chooseAstClosestToLocation(chpl::Context* chapel,
+                           const chpl::uast::AstNode* astOld,
+                           const chpl::uast::AstNode* astNew,
+                           chpl::Location loc) {
+  using namespace chpl::parsing;
+  CHPL_ASSERT(astNew);
+  CHPL_ASSERT(astOld != astNew);
+
+  auto locOld = astOld ? locateAst(chapel, astOld) : chpl::Location();
+  auto locNew = locateAst(chapel, astNew);
+  if (!locNew.contains(loc)) return astOld;
+  if (!astOld) return astNew;
+
+  // Prefer the location furthest towards the end of the range.
+  auto ret = (locNew > locOld) ? astNew : astOld;
+  return ret;
+}
+
+static const chpl::uast::AstNode*
+astAtLocation(Server* ctx, chpl::Location loc) {
+  const auto& m = ctx->withChapel(mapLinesToIdsInModuleQuery,
+                                  loc.path());
+  auto it = m.find(loc.firstLine());
+  if (it == m.end()) return nullptr;
+
+  ctx->trace("Found '%zu' AST nodes on line '%d'\n",
+             it->second.size(),
+             loc.firstLine());
+
+  const chpl::uast::AstNode* ret = nullptr;
+  for (auto id : it->second) {
+    auto last = ret;
+    ctx->withChapel([&](auto chapel) {
+      auto ast = chpl::parsing::idToAst(chapel, id);
+
+      if (ctx->isLogTrace()) {
+        ctx->trace("Considering '%s' and '%s'\n",
+                   ctx->fmt(ret).c_str(),
+                   ctx->fmt(ast).c_str());
+      }
+
+      ret = chooseAstClosestToLocation(chapel, ret, ast, loc);
+
+      if (ctx->isLogTrace()) {
+        ctx->trace("Chose '%s'\n", ctx->fmt(ret).c_str());
+      }
+
+    });
+    if (ret == last) break;
+  }
+
+  return ret;
+}
+
+static const chpl::uast::Module*
+closestModule(chpl::Context* chapel, const chpl::uast::AstNode* ast) {
+  if (auto mod = ast->toModule()) return mod;
+  if (auto id = chpl::parsing::idToParentModule(chapel, ast->id())) {
+    auto mod = chpl::parsing::idToAst(chapel, id);
+    return mod->toModule();
+  }
+  return nullptr;
+}
+
+static const chpl::resolution::ResolutionResultByPostorderID&
+resolveModule(Server* ctx, chpl::ID idMod) {
+  return ctx->withChapel(chpl::resolution::resolveModule, idMod);
+}
+
+static std::vector<chpl::ID>
+sourceAstToIds(Server* ctx, const chpl::uast::AstNode* ast) {
+  std::vector<chpl::ID> ret;
+
+  if (!ast) return ret;
+
+  auto mod = ctx->withChapel(closestModule, ast);
+  CHPL_ASSERT(mod);
+
+  if (mod == ast) CHPLDEF_TODO();
+
+  ctx->trace("Target '%s' in containing module '%s'\n",
+             ctx->fmt(ast).c_str(),
+             ctx->fmt(mod).c_str());
+
+  auto& rr = resolveModule(ctx, mod->id());
+  auto& re = rr.byAst(ast);
+
+  if (re.isBuiltin()) CHPLDEF_TODO();
+
+  ctx->trace("Encountered '%zu' errors resolving\n",
+             ctx->errorHandler()->numErrors());
```

Comment:
`numErrors` isn't reset, so this will print a running total of the number of errors, not just the errors from this particular resolution, right?